### ISSUE TITLE
Inconsistent Behaviour of `Optional[...]` together with `... is None` Check

### ISSUE THREAD

**Issue by hf-kklein** [python/mypy#14036]
**Bug Report**
A  `is None` check on variable which is typed as optional sometimes works as expected, sometimes doesn't.
This bug is mainly about the inconsistent behaviour of mypy nad not about one of both possibilities.

**To Reproduce**
[MyPy Playground](https://mypy-play.net/?mypy=latest&python=3.10&gist=98d5dcc8f68efb59bce9dd1b66b6f620)

```python
import re
from typing import Match, Optional

pattern = re.compile(".")


def my_function_a(a_string_or_none: Optional[str]):
    if a_string_or_none is None:
        return
    # at this point a_string_or_none is definitely not None
    match: Match[str] = pattern.match(a_string_or_none) # problem: error message below
    # error: Incompatible types in assignment (expression has type "Optional[Match[str]]", variable has type "Match[str]")  [assignment]


def my_function_b(a_string_or_none: Optional[str]):
    if a_string_or_none is None:
        return
    # at this point a_string_or_none is definitely not None
    function_that_only_consumes_strings(a_string_or_none)  # no problem here


def function_that_only_consumes_strings(a_string: str):
    pass

```
The inconsistency is, that mypy reports an error in `my_function_a` and does not report an error in `my_function_b`.

**Expected Behavior**
Consistent behaviour: Both `my_function_a` and `my_function_b` should report an error or both report no error.
If there is an error in line 11
> main.py:11: error: Incompatible types in assignment (expression has type "Optional[Match[str]]", variable has type "Match[str]")

Then there should also be an error in line 19.
> ⚠ (this error does not happen but I expected it to happen) Argument 1 to "function_that_only_consumes_strings" has incompatible type "Optional[str]"; expected "str"

But if there is no error in line 11, then there should also be no error in line 19 (which is the case right now)

**Actual Behavior**
The inconsistency is, that mypy reports an error in `my_function_a` and does not report an error in `my_function_b`.

**Your Environment**

- Mypy version used: 0.990
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10.0
